File: c-primes/primes.c
#define _POSIX_C_SOURCE 200809L
#include "primes.h"
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdio.h>
#include <stdint.h>
#include <unistd.h>

/* Integer square root using Newton's method */
size_t isqrt(size_t n) {
    if (n == 0) return 0;
    
    size_t x = (size_t)sqrt((double)n);
    
    /* Refine the estimate */
    while (x > 0 && x * x > n) {
        x--;
    }
    while ((x + 1) <= n / (x + 1)) {
        x++;
    }
    
    return x;
}

/* Estimate the number of primes up to n using Prime Number Theorem */
size_t estimate_prime_count(size_t n) {
    if (n <= 2) return 1;
    
    double ln_n = log((double)n);
    size_t estimated;
    
    if (ln_n > 1.1) {
        estimated = (size_t)((double)n / (ln_n - 1.1));
    } else {
        estimated = n;
    }
    
    return estimated > 0 ? estimated : 1;
}

/* Allocate and initialize a prime_result_t */
static prime_result_t create_result(size_t capacity) {
    prime_result_t result;
    result.capacity = capacity > 0 ? capacity : 1;
    result.primes = (size_t *)malloc(result.capacity * sizeof(size_t));
    result.count = 0;
    result.error = PRIME_OK;
    result.error_msg = NULL;
    return result;
}

/* Add a prime to the result, resizing if necessary */
static void add_prime(prime_result_t *result, size_t prime) {
    if (result->count >= result->capacity) {
        size_t new_capacity = result->capacity * 2;
        size_t *new_primes = (size_t *)realloc(result->primes, new_capacity * sizeof(size_t));
        if (new_primes) {
            result->primes = new_primes;
            result->capacity = new_capacity;
        }
    }
    if (result->count < result->capacity) {
        result->primes[result->count++] = prime;
    }
}



/* Classic Sieve of Eratosthenes (odd-only)
 * Best for n < 1,000,000
 */
prime_result_t sieve_of_eratosthenes(size_t n, progress_callback_t progress) {
    if (n <= 2) {
        prime_result_t result = create_result(0);
        return result;
    }
    
    if (n <= 3) {
        prime_result_t result = create_result(1);
        add_prime(&result, 2);
        return result;
    }
    
    /* Odd-only sieve: index i represents number 2*i + 3 */
    size_t sieve_size = (n - 3) / 2 + 1;
    if (sieve_size < 1) sieve_size = 1;
    
    bool *sieve = (bool *)calloc(sieve_size, sizeof(bool));
    if (!sieve) {
        prime_result_t result = create_result(0);
        result.error = PRIME_ERR_MEMORY_ALLOCATION;
        result.error_msg = strdup("Failed to allocate sieve array");
        return result;
    }
    
    /* Initialize all to true */
    for (size_t i = 0; i < sieve_size; i++) {
        sieve[i] = true;
    }
    
    size_t limit = isqrt(n);
    size_t current = 3;
    
    while (current <= limit) {
        size_t idx = (current - 3) / 2;
        if (idx < sieve_size && sieve[idx]) {
            unsigned long long squared = (unsigned long long)current * (unsigned long long)current;
            if (squared >= (unsigned long long)n) break;
            
            size_t start_idx = (size_t)((squared - 3) / 2);
            size_t step = current;
            
            for (size_t j = start_idx; j < sieve_size; j += step) {
                sieve[j] = false;
            }
        }
        current += 2;
    }
    
    /* Extract primes */
    prime_result_t result = create_result(estimate_prime_count(n));
    add_prime(&result, 2);
    
    for (size_t i = 0; i < sieve_size; i++) {
        if (sieve[i]) {
            size_t prime = 2 * i + 3;
            if (prime < n) {
                add_prime(&result, prime);
            }
        }
    }
    
    free(sieve);
    
    /* Invoke progress callback */
    if (progress) {
        progress(1);
    }
    
    return result;
}

/* Process a single segment using odd-only sieve
 * Shared helper for segmented and parallel sieves
 */
static void sieve_segment_odd_only(
    size_t low, size_t high,
    const size_t *base_primes_odd, size_t base_count,
    bool *is_prime, size_t buffer_size,
    size_t *segment_primes, size_t *seg_count, size_t seg_capacity
) {
    *seg_count = 0;
    
    /* Handle prime 2 if in this segment */
    if (low <= 2 && high > 2) {
        if (*seg_count < seg_capacity) {
            segment_primes[(*seg_count)++] = 2;
        }
    }
    
    /* Find first odd number >= low */
    size_t odd_low = low < 3 ? 3 : low;
    if (odd_low % 2 == 0) odd_low++;
    
    if (odd_low >= high) return;
    
    size_t seg_len = (high - odd_low + 1) / 2;
    if (seg_len == 0) return;
    if (seg_len > buffer_size) seg_len = buffer_size;
    
    /* Reset buffer */
    for (size_t i = 0; i < seg_len; i++) {
        is_prime[i] = true;
    }
    
    /* Sieve with base primes */
    for (size_t i = 0; i < base_count; i++) {
        size_t p = base_primes_odd[i];
        
        /* Find first odd multiple of p in [odd_low, high) */
        size_t start = (low + p - 1) / p * p;
        if (start < p * p) start = p * p;
        if (start % 2 == 0) start += p;
        
        if (start >= high) continue;
        
        size_t adjusted_start = (start - odd_low) / 2;
        size_t step = p;
        
        for (size_t j = adjusted_start; j < seg_len; j += step) {
            is_prime[j] = false;
        }
    }
    
    /* Extract primes from segment */
    for (size_t i = 0; i < seg_len; i++) {
        if (is_prime[i]) {
            size_t prime = odd_low + 2 * i;
            if (prime < high && *seg_count < seg_capacity) {
                segment_primes[(*seg_count)++] = prime;
            }
        }
    }
}

/* Segmented Sieve of Eratosthenes (odd-only)
 * Best for n >= 1,000,000
 */
prime_result_t segmented_sieve(size_t n, size_t segment_size, progress_callback_t progress) {
    if (n <= 2) {
        prime_result_t result = create_result(0);
        return result;
    }
    
    /* Validate segment size */
    size_t segments = n / segment_size + (n % segment_size > 0 ? 1 : 0);
    if (segments == 0) segments = 1;
    
    /* Generate base primes up to sqrt(n) */
    size_t base_limit = isqrt(n);
    prime_result_t base_result = sieve_of_eratosthenes(base_limit + 1, NULL);
    if (base_result.error != PRIME_OK) {
        prime_result_t result = create_result(0);
        result.error = base_result.error;
        result.error_msg = base_result.error_msg;
        return result;
    }
    
    /* Extract odd base primes only */
    size_t *base_primes_odd = NULL;
    size_t base_odd_count = 0;
    
    if (base_result.count > 0) {
        base_primes_odd = (size_t *)malloc(base_result.count * sizeof(size_t));
        if (!base_primes_odd) {
            free_prime_result(&base_result);
            prime_result_t result = create_result(0);
            result.error = PRIME_ERR_MEMORY_ALLOCATION;
            result.error_msg = strdup("Failed to allocate base primes array");
            return result;
        }
        
        for (size_t i = 0; i < base_result.count; i++) {
            if (base_result.primes[i] > 2) {
                base_primes_odd[base_odd_count++] = base_result.primes[i];
            }
        }
    }
    
    free_prime_result(&base_result);
    
    /* Allocate result */
    prime_result_t result = create_result(estimate_prime_count(n));
    
    /* Reusable buffers for segments */
    size_t buffer_size = segment_size / 2 + 1;
    bool *is_prime = (bool *)malloc(buffer_size * sizeof(bool));
    if (!is_prime) {
        free(base_primes_odd);
        result.error = PRIME_ERR_MEMORY_ALLOCATION;
        result.error_msg = strdup("Failed to allocate segment buffer");
        return result;
    }
    
    size_t *seg_primes = (size_t *)malloc(segment_size * sizeof(size_t));
    if (!seg_primes) {
        free(is_prime);
        free(base_primes_odd);
        result.error = PRIME_ERR_MEMORY_ALLOCATION;
        result.error_msg = strdup("Failed to allocate segment primes array");
        return result;
    }
    
    /* Process segments */
    for (size_t seg_idx = 0; seg_idx < segments; seg_idx++) {
        size_t low = seg_idx * segment_size;
        size_t high = low + segment_size;
        if (high > n) high = n;
        
        if (high <= 2) continue;
        
        size_t seg_count = 0;
        sieve_segment_odd_only(low, high, base_primes_odd, base_odd_count,
                              is_prime, buffer_size, seg_primes, &seg_count, segment_size);
        
        /* Add to result */
        for (size_t i = 0; i < seg_count && result.count < result.capacity; i++) {
            add_prime(&result, seg_primes[i]);
        }
        
        /* Progress callback */
        if (progress) {
            progress(1);
        }
    }
    
    free(seg_primes);
    free(is_prime);
    free(base_primes_odd);
    
    return result;
}

/* Worker thread function for parallel segmented sieve
 * Buffers (worker_primes, is_prime_buffer) are pre-allocated by the caller
 */
static void *worker_thread(void *arg) {
    worker_context_t *ctx = (worker_context_t *)arg;
    
    ctx->worker_count = 0;
    size_t buffer_size = ctx->segment_size / 2 + 1;
    
    for (size_t seg_idx = ctx->start_seg; seg_idx < ctx->end_seg; seg_idx++) {
        size_t low = seg_idx * ctx->segment_size;
        size_t high = low + ctx->segment_size;
        if (high > ctx->n) high = ctx->n;
        
        if (high <= 2) continue;
        
        size_t seg_count = 0;
        sieve_segment_odd_only(low, high, ctx->base_primes_odd, ctx->base_primes_count,
                              ctx->is_prime_buffer, buffer_size,
                              ctx->worker_primes + ctx->worker_count, &seg_count, 
                              ctx->worker_capacity - ctx->worker_count);
        
        ctx->worker_count += seg_count;
        
        if (ctx->progress) {
            ctx->progress(1);
        }
    }
    
    ctx->error = PRIME_OK;
    return NULL;
}

/* Parallel Segmented Sieve (odd-only)
 * Best for n >= 5,000,000
 */
prime_result_t parallel_segmented_sieve(size_t n, size_t workers, size_t segment_size, progress_callback_t progress) {
    if (n <= 2) {
        prime_result_t result = create_result(0);
        return result;
    }
    
    /* Auto-detect available parallelism if workers not specified */
    if (workers == 0) {
#ifdef _SC_NPROCESSORS_ONLN
        long ncpu = sysconf(_SC_NPROCESSORS_ONLN);
        workers = (ncpu > 0) ? (size_t)ncpu : 4;
#else
        workers = 4;
#endif
    }
    if (workers > MAX_WORKERS) workers = MAX_WORKERS;
    
    /* Validate segment size */
    size_t segments = n / segment_size + (n % segment_size > 0 ? 1 : 0);
    if (segments == 0) segments = 1;
    
    /* Adjust workers to not exceed segments */
    if (workers > segments) workers = segments;
    
    /* Generate base primes up to sqrt(n) */
    size_t base_limit = isqrt(n);
    prime_result_t base_result = sieve_of_eratosthenes(base_limit + 1, NULL);
    if (base_result.error != PRIME_OK) {
        prime_result_t result = create_result(0);
        result.error = base_result.error;
        result.error_msg = base_result.error_msg;
        return result;
    }
    
    /* Extract odd base primes only */
    size_t *base_primes_odd = NULL;
    size_t base_odd_count = 0;
    
    if (base_result.count > 0) {
        base_primes_odd = (size_t *)malloc(base_result.count * sizeof(size_t));
        if (!base_primes_odd) {
            free_prime_result(&base_result);
            prime_result_t result = create_result(0);
            result.error = PRIME_ERR_MEMORY_ALLOCATION;
            result.error_msg = strdup("Failed to allocate base primes array");
            return result;
        }
        
        for (size_t i = 0; i < base_result.count; i++) {
            if (base_result.primes[i] > 2) {
                base_primes_odd[base_odd_count++] = base_result.primes[i];
            }
        }
    }
    
    free_prime_result(&base_result);
    
    /* Calculate chunk size per worker */
    size_t chunk_size = segments / workers;
    if (segments % workers > 0) chunk_size++;
    
    /* Create worker contexts and threads */
    worker_context_t *contexts = (worker_context_t *)malloc(workers * sizeof(worker_context_t));
    pthread_t *threads = (pthread_t *)malloc(workers * sizeof(pthread_t));
    
    if (!contexts || !threads) {
        free(base_primes_odd);
        prime_result_t result = create_result(0);
        result.error = PRIME_ERR_MEMORY_ALLOCATION;
        result.error_msg = strdup("Failed to allocate worker structures");
        return result;
    }
    
    /* Pre-allocate worker buffers in main thread */
    size_t buffer_size = segment_size / 2 + 1;
    for (size_t w = 0; w < workers; w++) {
        size_t segs_for_worker = chunk_size;
        if (w * chunk_size + chunk_size > segments) {
            segs_for_worker = segments - w * chunk_size;
        }
        size_t worker_capacity = segs_for_worker * segment_size;
        
        contexts[w].worker_primes = (size_t *)malloc(worker_capacity * sizeof(size_t));
        contexts[w].is_prime_buffer = (bool *)malloc(buffer_size * sizeof(bool));
        
        if (!contexts[w].worker_primes || !contexts[w].is_prime_buffer) {
            /* Clean up on allocation failure */
            for (size_t k = 0; k <= w; k++) {
                free(contexts[k].worker_primes);
                free(contexts[k].is_prime_buffer);
            }
            free(contexts);
            free(threads);
            free(base_primes_odd);
            prime_result_t result = create_result(0);
            result.error = PRIME_ERR_MEMORY_ALLOCATION;
            result.error_msg = strdup("Failed to allocate worker buffers");
            return result;
        }
        
        contexts[w].worker_capacity = worker_capacity;
    }
    
    for (size_t w = 0; w < workers; w++) {
        contexts[w].n = n;
        contexts[w].start_seg = w * chunk_size;
        contexts[w].end_seg = contexts[w].start_seg + chunk_size;
        if (contexts[w].end_seg > segments) contexts[w].end_seg = segments;
        contexts[w].segment_size = segment_size;
        contexts[w].base_primes_odd = base_primes_odd;
        contexts[w].base_primes_count = base_odd_count;
        contexts[w].worker_count = 0;
        contexts[w].progress = progress;
        contexts[w].error = PRIME_OK;
        contexts[w].error_msg = NULL;
        
        if (contexts[w].start_seg >= segments) {
            contexts[w].end_seg = contexts[w].start_seg;
        }
    }
    
    /* Launch worker threads */
    size_t launched = 0;
    for (size_t w = 0; w < workers; w++) {
        if (contexts[w].start_seg >= contexts[w].end_seg) continue;
        
        int rc = pthread_create(&threads[launched], NULL, worker_thread, &contexts[w]);
        if (rc != 0) {
            prime_result_t result = create_result(0);
            result.error = PRIME_ERR_WORKER_THREAD_PANIC;
            char msg[64];
            snprintf(msg, sizeof(msg), "pthread_create failed with code %d", rc);
            result.error_msg = strdup(msg);
            
            /* Clean up */
            for (size_t k = 0; k < launched; k++) {
                pthread_join(threads[k], NULL);
            }
            for (size_t k = 0; k < workers; k++) {
                free(contexts[k].worker_primes);
                free(contexts[k].is_prime_buffer);
            }
            free(contexts);
            free(threads);
            free(base_primes_odd);
            return result;
        }
        launched++;
    }
    
    /* Wait for all workers */
    for (size_t w = 0; w < launched; w++) {
        pthread_join(threads[w], NULL);
    }
    
    /* Collect results from workers */
    prime_result_t result = create_result(estimate_prime_count(n));
    
    for (size_t w = 0; w < workers; w++) {
        if (contexts[w].error != PRIME_OK) {
            result.error = contexts[w].error;
            result.error_msg = contexts[w].error_msg;
            break;
        }
        
        for (size_t i = 0; i < contexts[w].worker_count; i++) {
            add_prime(&result, contexts[w].worker_primes[i]);
        }
    }
    
    /* Free pre-allocated worker buffers */
    for (size_t w = 0; w < workers; w++) {
        free(contexts[w].worker_primes);
        free(contexts[w].is_prime_buffer);
    }
    
    free(contexts);
    free(threads);
    free(base_primes_odd);
    
    return result;
}

/* Auto-select algorithm based on n */
prime_result_t generate_primes(size_t n, bool parallel, size_t workers, size_t segment_size, progress_callback_t progress) {
    if (n <= 2) {
        prime_result_t result = create_result(0);
        return result;
    }
    
    /* Check MAX_N */
    if (n > MAX_N) {
        prime_result_t result = create_result(0);
        result.error = PRIME_ERR_INVALID_INPUT;
        char msg[128];
        snprintf(msg, sizeof(msg), "n=%zu exceeds maximum supported value %llu (1 quadrillion)", n, (unsigned long long)MAX_N);
        result.error_msg = strdup(msg);
        return result;
    }
    
    /* Validate workers */
    if (workers == 0) {
        workers = 4; /* Default */
    }
    if (workers > MAX_WORKERS) workers = MAX_WORKERS;
    
    /* Validate segment size */
    if (segment_size == 0) {
        segment_size = DEFAULT_SEGMENT_SIZE;
    }
    
    /* Select algorithm */
    if (parallel && n >= PARALLEL_THRESHOLD) {
        return parallel_segmented_sieve(n, workers, segment_size, progress);
    } else if (n >= DEFAULT_SEGMENT_SIZE) {
        return segmented_sieve(n, segment_size, progress);
    } else {
        return sieve_of_eratosthenes(n, progress);
    }
}

/* Free prime_result_t resources */
void free_prime_result(prime_result_t *result) {
    if (result->primes) {
        free(result->primes);
        result->primes = NULL;
    }
    if (result->error_msg) {
        free(result->error_msg);
        result->error_msg = NULL;
    }
    result->count = 0;
    result->capacity = 0;
}

/* Validate segment size */
prime_error_t validate_segment_size(size_t n, size_t segment_size, char **error_msg) {
    if (segment_size == 0) {
        if (error_msg) {
            *error_msg = strdup("segment_size cannot be zero");
        }
        return PRIME_ERR_INVALID_INPUT;
    }
    
    size_t segments = n / segment_size + (n % segment_size > 0 ? 1 : 0);
    if (segments > 0 && segment_size > UINTPTR_MAX / segments) {
        if (error_msg) {
            *error_msg = strdup("segment_size would cause overflow in calculations");
        }
        return PRIME_ERR_INVALID_INPUT;
    }
    
    return PRIME_OK;
}

/* Validate workers parameter */
prime_error_t validate_workers(size_t workers, char **error_msg) {
    if (workers == 0) {
        if (error_msg) {
            *error_msg = strdup("workers must be >= 1");
        }
        return PRIME_ERR_INVALID_INPUT;
    }
    
    if (workers > MAX_WORKERS) {
        if (error_msg) {
            char msg[64];
            snprintf(msg, sizeof(msg), "workers=%zu exceeds maximum %d", workers, MAX_WORKERS);
            *error_msg = strdup(msg);
        }
        return PRIME_ERR_INVALID_INPUT;
    }
    
    return PRIME_OK;
}

File: golang-primes/prime/primes.go
package prime

import (
	"bytes"
	"errors"
	"fmt"
	"math"
	"runtime"
	"sync"
	"sync/atomic"
)

const (
	DefaultSegmentSize = 1_000_000
	ParallelThreshold  = 5_000_000

	// MaxN is the maximum supported input size (1 quadrillion)
	MaxN = 1_000_000_000_000
)

// ValidateError represents input validation errors
type ValidateError struct {
	msg string
}

func (e *ValidateError) Error() string {
	return e.msg
}

func (e *ValidateError) String() string {
	return e.msg
}

// NewValidateError creates a new ValidateError with the given message.
func NewValidateError(msg string) error {
	return &ValidateError{msg: msg}
}

func isValidateError(err error) bool {
	var ve *ValidateError
	return errors.As(err, &ve)
}

// validateN validates that n does not exceed the maximum supported input size.
func validateN(n int) error {
	if n < 0 {
		return NewValidateError(fmt.Sprintf("n=%d is invalid: must be >= 0", n))
	}
	if n > MaxN {
		return NewValidateError(fmt.Sprintf("n=%d exceeds maximum supported value %d (1 quadrillion). Generating primes above this limit would require impractical computation time.", n, MaxN))
	}
	return nil
}

// validateWorkers validates the workers parameter.
func validateWorkers(workers int) error {
	if workers < 0 {
		return NewValidateError(fmt.Sprintf("workers=%d is invalid: must be >= 0", workers))
	}
	if workers > 1024 {
		return NewValidateError(fmt.Sprintf("workers=%d exceeds maximum allowed value of 1024", workers))
	}
	return nil
}

// validateSegmentSize validates that segment_size is non-zero and reasonable.
func validateSegmentSize(segmentSize int) error {
	if segmentSize == 0 {
		return NewValidateError("segment_size cannot be zero")
	}
	if segmentSize < 0 {
		return NewValidateError(fmt.Sprintf("segment_size=%d is invalid: must be >= 0", segmentSize))
	}
	return nil
}

// sieveSegmentOddOnly processes a single segment using odd-only sieve.
// Shared helper used by both sequential and parallel segmented sieves.
//
// low/high define the segment range [low, high).
// basePrimes are odd primes up to sqrt(n) (excludes 2).
// isPrime is a reusable buffer (at least (high-low)/2 bytes).
// Returns primes found in [max(low,2), high).
func sieveSegmentOddOnly(low, high int, basePrimes []int, isPrime []byte) []int {
	var primes []int

	// Handle the prime 2 if it falls in this segment
	if low <= 2 && high > 2 {
		primes = append(primes, 2)
	}

	// Odd-only sieve: index i represents number oddLow + 2*i
	oddLow := low
	if oddLow < 3 {
		oddLow = 3
	}
	if oddLow%2 == 0 {
		oddLow++
	}
	if oddLow >= high {
		return primes
	}

	segLen := (high - oddLow + 1) / 2 // count of odd numbers in [oddLow, high)
	if segLen <= 0 {
		return primes
	}

	for i := 0; i < segLen; i++ {
		isPrime[i] = 1
	}

	for _, p := range basePrimes {
		// Find first odd multiple of p in [oddLow, high)
		start := ((low + p - 1) / p) * p
		if start < p*p {
			start = p * p
		}
		if start%2 == 0 {
			start += p
		}

		if start >= high {
			continue
		}

		// Map to index in odd-only array
		adjustedStart := (start - oddLow) / 2
		step := p // step in index space = p (each index step = 2 numbers)
		for j := adjustedStart; j < segLen; j += step {
			isPrime[j] = 0
		}
	}

	// Extract primes using bytes.IndexByte for SIMD-optimized search
	data := isPrime[:segLen]
	idx := 0
	for {
		pos := bytes.IndexByte(data[idx:], 1)
		if pos == -1 {
			break
		}
		idx += pos
		primes = append(primes, oddLow+2*idx)
		idx++
		if idx >= segLen {
			break
		}
	}

	return primes
}

func SieveOfEratosthenes(n int) ([]int, error) {
	if err := validateN(n); err != nil {
		return nil, err
	}

	if n <= 2 {
		return []int{}, nil
	}

	if n <= 3 {
		return []int{2}, nil
	}

	// Odd-only sieve: index i represents number 2*i + 3
	sieveSize := (n - 3 + 1) / 2 // count of odd numbers in [3, n)
	sieve := make([]byte, sieveSize)
	for i := 0; i < sieveSize; i++ {
		sieve[i] = 1
	}

	limit := int(math.Sqrt(float64(n)))
	for current := 3; current <= limit; current += 2 {
		idx := (current - 3) / 2
		if sieve[idx] == 1 {
			// Mark multiples of current starting at current*current
			startIdx := (current*current - 3) / 2
			step := current // step in index space
			for j := startIdx; j < sieveSize; j += step {
				sieve[j] = 0
			}
		}
	}

	// Pre-allocate with prime counting estimate
	estimated := int(float64(n) / math.Log(float64(n)) * 1.1)
	primes := make([]int, 0, estimated)
	primes = append(primes, 2)

	// Extract primes using bytes.IndexByte for SIMD-optimized search
	idx := 0
	for {
		pos := bytes.IndexByte(sieve[idx:], 1)
		if pos == -1 {
			break
		}
		idx += pos
		primes = append(primes, 2*idx+3)
		idx++
		if idx >= sieveSize {
			break
		}
	}

	return primes, nil
}

// PrimeGenError is returned when prime generation fails.
type PrimeGenError struct {
	msg string
}

func (e *PrimeGenError) Error() string {
	return e.msg
}

// WorkerThreadPanic creates an error for worker thread panics.
func WorkerThreadPanic(msg string) error {
	return &PrimeGenError{msg: fmt.Sprintf("worker thread panicked: %s", msg)}
}

// InvalidInput creates an error for invalid input parameters.
func InvalidInput(msg string) error {
	return &PrimeGenError{msg: fmt.Sprintf("invalid input: %s", msg)}
}

// IsValidateError returns true if the error is a validation error.
func IsValidateError(err error) bool {
	return isValidateError(err)
}

func SegmentedSieve(n int, segmentSize int, progress func(int)) ([]int, error) {
	if err := validateN(n); err != nil {
		return nil, err
	}
	if err := validateSegmentSize(segmentSize); err != nil {
		return nil, err
	}

	if n <= 2 {
		return []int{}, nil
	}
	if segmentSize == 0 {
		segmentSize = DefaultSegmentSize
	}

	baseLimit := int(math.Sqrt(float64(n)))
	allBasePrimes, err := SieveOfEratosthenes(baseLimit + 1)
	if err != nil {
		return nil, err
	}
	basePrimesOdd := allBasePrimes[1:]

	segments := (n + segmentSize - 1) / segmentSize
	estimated := int(float64(n) / math.Log(float64(n)) * 1.1)
	primes := make([]int, 0, estimated)

	// Reusable buffer for segments -- allocate once to max segment size
	isPrime := make([]byte, segmentSize)

	for segIdx := 0; segIdx < segments; segIdx++ {
		low := segIdx * segmentSize
		high := low + segmentSize
		if high > n {
			high = n
		}

		if high <= 2 {
			if progress != nil {
				progress(1)
			}
			continue
		}

		segPrimes := sieveSegmentOddOnly(low, high, basePrimesOdd, isPrime)
		primes = append(primes, segPrimes...)

		if progress != nil {
			progress(1)
		}
	}

	return primes, nil
}

type segmentWork struct {
	segIdx int
	low    int
	high   int
}

type segmentResult struct {
	segIdx int
	primes []int
}

func workerProcessSegment(
	workChan <-chan segmentWork,
	resultsChan chan<- segmentResult,
	basePrimes []int,
	segmentSize int,
	wg *sync.WaitGroup,
	completedSegments *int64,
) {
	defer wg.Done()
	// Each worker gets its own reusable buffer
	isPrime := make([]byte, segmentSize)

	for work := range workChan {
		if work.high <= 2 {
			atomic.AddInt64(completedSegments, 1)
			resultsChan <- segmentResult{segIdx: work.segIdx, primes: nil}
			continue
		}

		segPrimes := sieveSegmentOddOnly(work.low, work.high, basePrimes, isPrime)
		atomic.AddInt64(completedSegments, 1)

		resultsChan <- segmentResult{
			segIdx: work.segIdx,
			primes: segPrimes,
		}
	}
}

func ParallelSegmentedSieve(n int, workers, segmentSize int, progress func(int)) ([]int, error) {
	if err := validateN(n); err != nil {
		return nil, err
	}
	if segmentSize == 0 {
		segmentSize = DefaultSegmentSize
	} else if err := validateSegmentSize(segmentSize); err != nil {
		return nil, err
	}
	if err := validateWorkers(workers); err != nil {
		return nil, err
	}

	if n <= 2 {
		return []int{}, nil
	}
	if workers <= 0 {
		workers = runtime.NumCPU()
	}

	baseLimit := int(math.Sqrt(float64(n)))
	allBasePrimes, err := SieveOfEratosthenes(baseLimit + 1)
	if err != nil {
		return nil, err
	}
	basePrimesOdd := allBasePrimes[1:]

	segments := (n + segmentSize - 1) / segmentSize
	numWorkers := workers
	if numWorkers > segments {
		numWorkers = segments
	}

	// Bounded channels to limit memory usage
	workChan := make(chan segmentWork, numWorkers*2)
	resultsChan := make(chan segmentResult, numWorkers*2)
	var wg sync.WaitGroup
	var completedSegments int64

	// Start workers
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go workerProcessSegment(workChan, resultsChan, basePrimesOdd, segmentSize, &wg, &completedSegments)
	}

	// Producer: enqueue all segments
	go func() {
		for segIdx := 0; segIdx < segments; segIdx++ {
			low := segIdx * segmentSize
			high := low + segmentSize
			if high > n {
				high = n
			}

			workChan <- segmentWork{
				segIdx: segIdx,
				low:    low,
				high:   high,
			}
		}
		close(workChan)
	}()

	// Closer: wait for all workers then close results
	go func() {
		wg.Wait()
		close(resultsChan)
	}()

	// Progress monitor goroutine
	var progressDone chan struct{}
	if progress != nil {
		progressDone = make(chan struct{})
		go func() {
			defer close(progressDone)
			var lastSeen int64
			for {
				current := atomic.LoadInt64(&completedSegments)
				if current > lastSeen {
					delta := int(current - lastSeen)
					progress(delta)
					lastSeen = current
				}
				if current >= int64(segments) {
					return
				}
				runtime.Gosched()
			}
		}()
	}

	// Collect results indexed by segment for ordered reassembly
	results := make([][]int, segments)
	totalPrimes := 0
	for result := range resultsChan {
		results[result.segIdx] = result.primes
		totalPrimes += len(result.primes)
	}

	// Wait for progress monitor to finish
	if progressDone != nil {
		<-progressDone
	}

	// Flatten in segment order -- single allocation
	allPrimes := make([]int, 0, totalPrimes)
	for _, segPrimes := range results {
		allPrimes = append(allPrimes, segPrimes...)
	}

	return allPrimes, nil
}

func GeneratePrimes(n int, parallel bool, progress func(int)) ([]int, error) {
	if err := validateN(n); err != nil {
		return nil, err
	}

	if n <= 2 {
		return []int{}, nil
	}

	if parallel && n >= ParallelThreshold {
		return ParallelSegmentedSieve(n, 0, DefaultSegmentSize, progress)
	}

	if n >= DefaultSegmentSize {
		return SegmentedSieve(n, DefaultSegmentSize, progress)
	}

	return SieveOfEratosthenes(n)
}

// ProgressTracker provides thread-safe progress tracking using atomics.
type ProgressTracker struct {
	total     int64
	completed int64
}

func NewProgressTracker(total int64) *ProgressTracker {
	return &ProgressTracker{total: total}
}

func (p *ProgressTracker) AddCompleted(delta int64) {
	atomic.AddInt64(&p.completed, delta)
}

func (p *ProgressTracker) GetCompleted() int64 {
	return atomic.LoadInt64(&p.completed)
}

func (p *ProgressTracker) GetPercent() int {
	if p.total == 0 {
		return 100
	}
	return int(float64(atomic.LoadInt64(&p.completed)) / float64(p.total) * 100)
}
